_posts/2020-10-24-spring-boot-specification.md

```diff
@@ -0,0 +1,372 @@
+add more based on your requirement.
+
+Now, as we know criteria API allows us to write typesafe queries. So, the values that we provide must be of the type compatible with the type of our field. `QueryInput` takes the value as `String` type which means we will have to cast the values to a required type 
+before passing it to `CriteriaBuilder`. 
+
+```java
+private Object castToRequiredType(Class fieldType, String value) {
+  if(fieldType.isAssignableFrom(Double.class)){
+      return Double.valueOf(value);
+  }else if(fieldType.isAssignableFrom(Integer.class)){
+      return Integer.valueOf(value);
+  }else if(Enum.class.isAssignableFrom(fieldType)){
+      return Enum.valueOf(fieldType, value);
+  }
+  return null;
+}
+
+private Object castToRequiredType(Class fieldType, List<String> value) {
+  List<Object> lists = new ArrayList<>();
+  for (String s : value) {
+      lists.add(castToRequiredType(fieldType, s));
+  }
+  return lists;
+}
+```
+
+Finally, we add a function that will combine multiple Specifications:
+
+```java
+private Specification<Product> getSpecificationFromQuery(List<QueryInput> queryInput){
+    Specification<Product> specification = 
+                        where(createSpecification(queryInput.remove(0)));
+    for (QueryInput input : queryInput) {
+        if(input.isOptional()){
+            specification = specification.or(createSpecification(input));
+        }else {
+            specification = specification.and(createSpecification(input));
+        }
+    }
+    return specification;
+}
+```
+
+Let's try to fetch all the products belonging to the `MOBILE` and `TV APPLIANCE` category and whose prices are below 1000 using 
+our new shiny dynamic specification query generator.
+
+```java
+QueryInput categories = QueryInput.builder()
+         .field("category")
+         .operator(QueryOperator.IN)
+         .values(List.of(Category.MOBILE.name(), 
+                         Category.TV_APPLIANCES.name()))
+         .isOptional(false)
+         .build();
+
+QueryInput lowRange = QueryInput.builder()
+        .field("price")
+        .operator(QueryOperator.LT)
+        .value("1000")
+        .isOptional(false)
+        .build();
+
+List<QueryInput> queries = new ArrayList<>();
+queries.add(lowRange);
+queries.add(categories);
+
+productAdapter.getQueryResult(queries);
+```
+
+The above code snippets should do for most filter cases but there is still a lot of room for improvement. Such as allowing queries 
+based on nested entity properties (`manufacturingPlace.state`), limiting the fields on which want to allow filters, etc. 
+Consider this as an open-end problem.
+
+## Conclusion
+
+Specification provides us with a way to write reusable queries and also fluent APIs with which we can combine and build more
+sophisticated queries.
+
+One question that comes to mind is that if we can write any query with Specification then should we never use query methods? I would say no. If our entity has only 
```

Comment:
"Never" and in the next sentence "I would say no". That's double negation and hard to understand the first time